pkg/cluster/k8sres.go

```diff
@@ -669,6 +669,20 @@ func (c *Cluster) generateSingleUserSecret(namespace string, pgUser spec.PgUser)
 	return &secret
 }
 
+func (c *Cluster) shouldCreateLoadBalancerForService(role PostgresRole, spec *spec.PostgresSpec) bool {
+
+	// handle LB for the replica service separately if ReplicaLoadBalancer is set in the Postgres manifest
+	if role == Replica && spec.ReplicaLoadBalancer != nil {
+		return *spec.ReplicaLoadBalancer
+	}
+
+	// create a load balancer for the master service if either Postgres or operator manifests tell to do so
+	// if ReplicaLoadBalancer is unset and LB for master service is created, also create a balancer for replicas
+	return (spec.UseLoadBalancer != nil && *spec.UseLoadBalancer) ||
+		(spec.UseLoadBalancer == nil && c.OpConfig.EnableLoadBalancer)
```

Comment:
It took me a while to understand this logic as it's written.

The cluster gets a replica load balancer according to:

1. the value of `replicaLoadBalancer` if specified;
2. otherwise, the value of `useLoadBalancer` if specified;
3. otherwise, the value of `enableLoadBalancer` if specified;
4. otherwise, yes.

For the primary service, it's the same but skipping step 1.

This looks kind of like

```
coalesce(
  role == replica ? spec.ReplicaLoadBalancer : nil,
  spec.UseLoadBalancer,
  c.OpConfig.EnableLoadBalancer,
  true,
)
```

…except that the `c.OpConfig.EnableLoadBalancer` field is not actually treated as nullable by representing it with a pointer to bool, but instead is represented as a non-pointer struct field defaulting to `true` on JSON deserialization, which is an odd but minor inconsistency.

This is unfortunately a bit unwieldy to express clearly in Go (oh wow), but maybe this is a bit more clear?

```
func Coalesce(values ...*bool) *bool {
	for _, value := range values {
		if value != nil {
			return value
		}
	}
	return nil
}

func Conditional(condition bool, value *bool) *bool {
	if condition {
		return value
	}
	return nil
}

…

default := true
Coalesce(
  Conditional(role == replica, spec.ReplicaLoadBalancer),
  spec.UseLoadBalancer,
  c.OpConfig.EnableLoadBalancer,
  &default,
)
```

This is kind of complicated and verbose and not necessarily better than how the logic is currently spelled out (and it's also unfortunate that `Coalesce` is of limited use since either `bool` has to be hardwired or casts would be needed to use it), so it may be best left as it's currently written.

I think a more explicit comment would help, though.

Reply:
By the way, the default should not be "true". It should be exactly what comes from opConfig EnableLoadBalancer, which is either true or false.

Reply:
Oops, yes, that final bit of code is wrong (even has the wrong type there); it would have to be like

```
Coalesce(
  Conditional(role == replica, spec.ReplicaLoadBalancer),
  spec.UseLoadBalancer,
  &c.OpConfig.EnableLoadBalancer,
)
```

(The default exists but is set when the struct is parsed from JSON)

Reply:
@mgomezch I already proposed one to have a function like your modification `Coalesce` and I like this idea, but in this case maybe just a few simple if conditions would be better? Like:

```go
if role == Replica && spec.ReplicaLoadBalancer != nil {
    return *spec.ReplicaLoadBalancer
}

if spec.UseLoadBalancer != nil {
    return *spec.UseLoadBalancer
}

return c.OpConfig.EnableLoadBalancer
```